# DS-SF-36 | Unit Project | 3 | Machine Learning Modeling and Executive Summary | Starter Code

In this project, you will perform a logistic regression on the admissions data we've been working with in Unit Project 1 and 2.  You will summarize and present your findings and the methods you used.

In [1]:
import os

import numpy as np
import pandas as pd
pd.set_option('display.max_rows', 10)
pd.set_option('display.max_columns', 10)
pd.set_option('display.notebook_repr_html', True)

import statsmodels.formula.api as smf

from sklearn import linear_model

In [2]:
df = pd.read_csv(os.path.join('..', '..', 'dataset', 'dataset-ucla-admissions.csv'))
df.dropna(inplace = True)

df

admit    gre   gpa  prestige
0        0  380.0  3.61       3.0
1        1  660.0  3.67       3.0
2        1  800.0  4.00       1.0
3        1  640.0  3.19       4.0
4        0  520.0  2.93       4.0
..     ...    ...   ...       ...
395      0  620.0  4.00       2.0
396      0  560.0  3.04       3.0
397      0  460.0  2.63       2.0
398      0  700.0  3.65       2.0
399      0  600.0  3.89       3.0

[397 rows x 4 columns]

## Part A.  Frequency Table

> ### Question 1.  Create a frequency table for `prestige` and whether an applicant was admitted.

In [3]:
df_frequency = pd.crosstab(index=df["prestige"], columns=df["admit"])
df_frequency

admit      0   1
prestige        
1.0       28  33
2.0       95  53
3.0       93  28
4.0       55  12

Answer: Above is a frequency table showing prestige and whether an applicant was admitted (0 = not admitted, 1= admitted)

## Part B.  Feature Engineering

> ### Question 2.  Create a one-hot encoding for `prestige`.

In [4]:
prestige_df = pd.get_dummies(df.prestige, prefix = 'Prestige')
prestige_df.rename(columns = {'Prestige_1.0': 'Prestige_1',
    'Prestige_2.0': 'Prestige_2',
    'Prestige_3.0': 'Prestige_3',
    'Prestige_4.0': 'Prestige_4'}, inplace = True)
df = df.join([prestige_df])
df.drop('prestige', axis = 1, inplace = True)
df

admit    gre   gpa  Prestige_1  Prestige_2  Prestige_3  Prestige_4
0        0  380.0  3.61           0           0           1           0
1        1  660.0  3.67           0           0           1           0
2        1  800.0  4.00           1           0           0           0
3        1  640.0  3.19           0           0           0           1
4        0  520.0  2.93           0           0           0           1
..     ...    ...   ...         ...         ...         ...         ...
395      0  620.0  4.00           0           1           0           0
396      0  560.0  3.04           0           0           1           0
397      0  460.0  2.63           0           1           0           0
398      0  700.0  3.65           0           1           0           0
399      0  600.0  3.89           0           0           1           0

[397 rows x 7 columns]

In [5]:
smf.ols(formula = 'admit ~ Prestige_2 + Prestige_3 + Prestige_4', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  admit   R-squared:                       0.064
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     8.899
Date:                Mon, 07 Aug 2017   Prob (F-statistic):           1.02e-05
Time:                        17:24:17   Log-Likelihood:                -246.67
No. Observations:                 397   AIC:                             501.3
Df Residuals:                     393   BIC:                             517.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.5410      0.058      9.333      0.000         0.427     0.655
Prestige_2    -0.1829      0.069     -2.655      0.008        -0.318    -0.047
Prestige_3    -0.3096      0.071     -4.355      0.000        -0.449    -0.170
Prestige_4    -0.3619      0.080     -4.517      0.000        -0.519    -0.204
==============================================================================
Omnibus:                      218.144   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.254
Skew:                           0.725   Prob(JB):                     3.69e-13
Kurtosis:                       1.834   Cond. No.                         6.10
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
smf.ols(formula = 'admit ~ Prestige_1 + Prestige_3 + Prestige_4', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  admit   R-squared:                       0.064
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     8.899
Date:                Mon, 07 Aug 2017   Prob (F-statistic):           1.02e-05
Time:                        17:24:17   Log-Likelihood:                -246.67
No. Observations:                 397   AIC:                             501.3
Df Residuals:                     393   BIC:                             517.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.3581      0.037      9.624      0.000         0.285     0.431
Prestige_1     0.1829      0.069      2.655      0.008         0.047     0.318
Prestige_3    -0.1267      0.055     -2.284      0.023        -0.236    -0.018
Prestige_4    -0.1790      0.067     -2.685      0.008        -0.310    -0.048
==============================================================================
Omnibus:                      218.144   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.254
Skew:                           0.725   Prob(JB):                     3.69e-13
Kurtosis:                       1.834   Cond. No.                         4.20
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [7]:
smf.ols(formula = 'admit ~ Prestige_1 + Prestige_2 + Prestige_4', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  admit   R-squared:                       0.064
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     8.899
Date:                Mon, 07 Aug 2017   Prob (F-statistic):           1.02e-05
Time:                        17:24:17   Log-Likelihood:                -246.67
No. Observations:                 397   AIC:                             501.3
Df Residuals:                     393   BIC:                             517.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.2314      0.041      5.623      0.000         0.150     0.312
Prestige_1     0.3096      0.071      4.355      0.000         0.170     0.449
Prestige_2     0.1267      0.055      2.284      0.023         0.018     0.236
Prestige_4    -0.0523      0.069     -0.759      0.449        -0.188     0.083
==============================================================================
Omnibus:                      218.144   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.254
Skew:                           0.725   Prob(JB):                     3.69e-13
Kurtosis:                       1.834   Cond. No.                         4.59
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [8]:
smf.ols(formula = 'admit ~ Prestige_1 + Prestige_2 + Prestige_3', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  admit   R-squared:                       0.064
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     8.899
Date:                Mon, 07 Aug 2017   Prob (F-statistic):           1.02e-05
Time:                        17:24:17   Log-Likelihood:                -246.67
No. Observations:                 397   AIC:                             501.3
Df Residuals:                     393   BIC:                             517.3
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept      0.1791      0.055      3.238      0.001         0.070     0.288
Prestige_1     0.3619      0.080      4.517      0.000         0.204     0.519
Prestige_2     0.1790      0.067      2.685      0.008         0.048     0.310
Prestige_3     0.0523      0.069      0.759      0.449        -0.083     0.188
==============================================================================
Omnibus:                      218.144   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               57.254
Skew:                           0.725   Prob(JB):                     3.69e-13
Kurtosis:                       1.834   Cond. No.                         5.87
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Answer: ran 4 regressions including 3 of the 4 binary variables created.  

> ### Question 3.  How many of these binary variables do we need for modeling?

Answer: 

> ### Question 4.  Why are we doing this?

Answer: TODO

> ### Question 5.  Add all these binary variables in the dataset and remove the now redundant `prestige` feature.

Answer: binary variables were added to the dataset and removed the redundant prestige feature under question 2. DF readout below confirms.

In [9]:
df

admit    gre   gpa  Prestige_1  Prestige_2  Prestige_3  Prestige_4
0        0  380.0  3.61           0           0           1           0
1        1  660.0  3.67           0           0           1           0
2        1  800.0  4.00           1           0           0           0
3        1  640.0  3.19           0           0           0           1
4        0  520.0  2.93           0           0           0           1
..     ...    ...   ...         ...         ...         ...         ...
395      0  620.0  4.00           0           1           0           0
396      0  560.0  3.04           0           0           1           0
397      0  460.0  2.63           0           1           0           0
398      0  700.0  3.65           0           1           0           0
399      0  600.0  3.89           0           0           1           0

[397 rows x 7 columns]

## Part C.  Hand calculating odds ratios

Let's develop our intuition about expected outcomes by hand calculating odds ratios.

> ### Question 6.  Create a frequency table for `prestige = 1` and whether an applicant was admitted.

In [10]:
df_frequency = pd.crosstab(index=df["Prestige_1"], columns=df["admit"])
df_frequency

admit         0   1
Prestige_1         
0           243  93
1            28  33

Answer: We are looking for when Prestige 1=1 and admit is 1.  Therefore of the 61 applicants from prestige=1 alma maters, 33 were admitted (33/61 = 54% of the prestige =1 applicants were admitted).

> ### Question 7.  Use the frequency table above to calculate the odds of being admitted to graduate school for applicants that attended the most prestigious undergraduate schools.

Answer: Odds of being admitted to graduate school for applicants whom attend the most prestigious undergraduate schools (prestige = 1) is 33/28 = 1.18
- 33 represents the number of prestige=1 applicants whom were admitted 
- 28 represents the number of prestige=1 applicants whom were not admitted

Another way to calculate odds is: probability of acceptance/(1-probability of acceptance):
    =(0.54)/(1-0.54)=1.17 ~ 1.18

> ### Question 8.  Now calculate the odds of admission for undergraduates who did not attend a #1 ranked college.

Answer: 0.38
From the frequency table in Question 6 above, there were 336 (=243+93) applicants whom did not attend a #1 ranked college.  The odds of admission can be calculated by dividing the number of admitted students (93) by the number of non-admitted students (243)
    = 93/243 = 0.38.  

> ### Question 9.  Finally, what's the odds ratio?

Answer: Odds ratio for undergraduates whom did not attend a #1 school is 93/243 (admitted vs not admitted)

The odds ratio of admittance if a student is from a prestige=1 undergraduate institution vs. admittance if they are NOT from a prestige = 1 institution can be calculated as follows:
    =1.18/.38=3.10

> ### Question 10.  Write this finding in a sentence.

Answer: Applicants whom attended a prestige=1 undergraduate institution are 3 times more likely to be admitted to graduate school based on the UCLA admittance data when compared to students whom did NOT attend a prestige=1 undergraduate institution.

> ### Question 11.  Use the frequency table above to calculate the odds of being admitted to graduate school for applicants that attended the least prestigious undergraduate schools.  Then calculate their odds ratio of being admitted to UCLA.  Finally, write this finding in a sentence.

In [11]:
df_frequency = pd.crosstab(index=df["Prestige_4"], columns=df["admit"])
df_frequency

admit         0    1
Prestige_4          
0           216  114
1            55   12

Answer: 
- Created a frequency table for the least prestigious undergraduate schools (prestige=4) above (aligns with data from original table).
- Odds of admittance for a student from a prestige=4 undergraduate school is 12/55 = .22
- Odds of admittance for a student from a more prestigious institution is 114/216=.53
- Odds ratio of a student from a prestige=4 undergraduate school being admitted to UCLA vs. a student from more prestigious institutions (prestige=1,2,or3) is .22/.53=.42 - they are less likely to be accepted (40% as likely).

## Part D. Analysis using `statsmodels`

> ### Question 12.  Fit a logistic regression model predicting admission into UCLA using `gre`, `gpa`, and the `prestige` of the undergraduate schools.  Use the highest prestige undergraduate schools as your reference point.

In [18]:
smf.ols(formula = 'admit ~ gre + gpa + Prestige_1 + Prestige_2 + Prestige_3', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  admit   R-squared:                       0.099
Model:                            OLS   Adj. R-squared:                  0.087
Method:                 Least Squares   F-statistic:                     8.594
Date:                Mon, 07 Aug 2017   Prob (F-statistic):           9.71e-08
Time:                        18:23:22   Log-Likelihood:                -239.02
No. Observations:                 397   AIC:                             490.0
Df Residuals:                     391   BIC:                             513.9
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
Intercept     -0.5617      0.208     -2.704      0.007        -0.970    -0.153
gre            0.0004      0.000      1.997      0.047      6.48e-06     0.001
gpa            0.1508      0.064      2.349      0.019         0.025     0.277
Prestige_1     0.3240      0.079      4.082      0.000         0.168     0.480
Prestige_2     0.1605      0.066      2.440      0.015         0.031     0.290
Prestige_3     0.0330      0.068      0.485      0.628        -0.101     0.167
==============================================================================
Omnibus:                      152.312   Durbin-Watson:                   1.946
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               50.314
Skew:                           0.678   Prob(JB):                     1.19e-11
Kurtosis:                       1.904   Cond. No.                     5.76e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.76e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [14]:
print np.exp(result.params)

NameError: name 'result' is not defined

> ### Question 13.  Print the model's summary results.

In [ ]:
# TODO

> ### Question 14.  What are the odds ratios of the different features and their 95% confidence intervals?

In [ ]:
# TODO

> ### Question 15.  Interpret the odds ratio for `prestige = 2`.

Answer: TODO

> ### Question 16.  Interpret the odds ratio of `gpa`.

Answer: TODO

> ### Question 17.  Assuming a student with a GRE of 800 and a GPA of 4.  What is his/her probability of admission  if he/she come from a tier-1, tier-2, tier-3, or tier-4 undergraduate school?

In [ ]:
# TODO

Answer: TODO

## Part E. Moving the model from `statsmodels` to `sklearn`

> ### Question 18.  Let's assume we are satisfied with our model.  Remodel it (same features) using `sklearn`.  When creating the logistic regression model with `LogisticRegression(C = 10 ** 2)`.

In [ ]:
# TODO

> ### Question 19.  What are the odds ratios for the different variables and how do they compare with the odds ratios calculated with `statsmodels`?

In [ ]:
# TODO

Answer: TODO

> ### Question 20.  Again, assuming a student with a GRE of 800 and a GPA of 4.  What is his/her probability of admission  if he/she come from a tier-1, tier-2, tier-3, or tier-4 undergraduate school?

In [ ]:
# TODO

Answer: TODO

## Part F.  Executive Summary

> ## Question 21.  Introduction
>
> Write a problem statement for this project.

Answer: TODO

> ## Question 22.  Dataset
>
> Write up a description of your data and any cleaning that was completed.

Answer: TODO

> ## Question 23.  Demo
>
> Provide a table that explains the data by admission status.

Answer: TODO

> ## Question 24.  Methods
>
> Write up the methods used in your analysis.

Answer: TODO

> ## Question 25.  Results
>
> Write up your results.

Answer: TODO

> ## Question 26.  Visuals
>
> Provide a table or visualization of these results.

Answer: TODO

> ## Question 27.  Discussion
>
> Write up your discussion and future steps.

Answer: TODO